### ISSUE TITLE
Queue operation in conditional execution context fails with "operation has been marked as not fetchable"

### ISSUE THREAD

**Issue by matthiasreisser** [tensorflow/tensorflow#4094]
Hey TensorFlow Community,
a while ago I wrote some code for doing gradient descent in a distributed environment. I adapted code from `tensorflow/tensorflow/python/training/sync_replicas_optimizer.py` and wrote my own `apply_gradients` method. Crucially, I included a conditional op, that either performed (distributed, synchronised) gradient descent or returned the `apply_gradients` op from the original optimizer. 
This code worked well when I used it with TensorFlow 0.8, but it broke with 0.9 and is still broken in my current installation from source (see below). The QueueRunner that is responsible for the synchronization op between the workers now fails with the error message "operation has been marked as not fetchable".
A bit of searching in the execution stack led me to the lines 
`if self._control_flow_context is not None:`
      `self._control_flow_context.AddOp(self)`
in `tensorflow/tensorflow/python/framework/ops.py`
from which I take that the error comes from the fact that the queue operations are placed inside the conditional execution context. 

Out of curiosity: What is the idea behind "fetchable" ops - and why should enqueuing and dequeueing be "dangerous to fetch"? 

Any help would be appreciated!

Matthias

### Environment info
Operating System: 
Distributor ID:	LinuxMint
Description:	Linux Mint 17.2 Rafaela
Release:	17.2
Codename:	rafaela

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):

cuda/lib64/libcudadevrt.a
cuda/lib64/libcudart.so -> libcudart.so.8.0
cuda/lib64/libcudart.so.8.0 -> libcudart.so.8.0.27
cuda/lib64/libcudart.so.8.0.27
cuda/lib64/libcudart_static.a
cuda/lib64/libcudnn.so
cuda/lib64/libcudnn.so.5
cuda/lib64/libcudnn.so.5.1.5
cuda/lib64/libcudnn_static.a
If installed from binary pip package, provide:

If installed from source, provide 

1. The commit hash (`git rev-parse HEAD`)
`57ff9d9be2f1faaba9598a3d99ef6c3af02342a4`

### What other attempted solutions have you tried?
Removing the conditional execution clause removes the error

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).

**Comment by mrry** [comment]
Any op created in a branch of a TensorFlow conditional or the body of a TensorFlow loop is marked as "non-fetchable", to prevent various programming errors. For example, there was once a bug where if you fetched the result of an op that was created in the *non-taken* branch of a conditional, it was possible to cause a crash (or perhaps a deadlock - I'm not 100% sure). Similarly, if you fetched the result of an op in a loop, the runtime would become confused because that tensor could have multiple values (one per iteration) at execution time. I'm not sure what your old code looked like, but it's possible that it was working by chance, given the way control flow is implemetned :).

In general, it should be possible to workaround this constraint by returning the ops that you want to run from the true-branch and false-branch functions (adding a `tf.no_op()` to the other side if necessary). Let us know if that isn't the case though.

**Comment by matthiasreisser** [comment]
Let me see if I understand correctly. In the True branch, I create a queue and define a QueueRunner object (which is stored and referenced later). This QueueRunner is then started by the tf.train.Supervisor() before starting the main training loop. Does this started QueueRunner then violate the assumption that the cond() branch is only a function? 

On another note, If I manipulate the condition to always evaluate to the "True" branch (which involves the queueing operations), the error still arises.

**Comment by mrry** [comment]
That's correct. Instead of creating the QueueRunner inside a branch of the cond, you should create it based on the result returned by the `tf.cond()` function. The fetchability of tensors is determined statically, before any optimizations take place, so the error will be raised even if the same branch is always taken.

**Comment by matthiasreisser** [comment]
Thanks for the help! Putting the QueueRunner logic outside the tf.cond() context, based on the result of the tf.cond() fixed the issue.

**Comment by danijar** [comment]
Let's say an op should be called conditionally using `tf.cond(predicate, lambda: op, lambda: tf.no_op())`, but also manually using `sess.run(op)`. The manual execution is prevented by the "marked as not fetchable" mechanism. Is this intended?

**Comment by mrry** [comment]
@danijar No, that sounds like a bug. Feel free to open another issue about this (or send a pull request with a fix)!

**Comment by vanpersie32** [comment]
@matthiasreisser  hi, I have similar problem. I create a queue runner object and  add it to collection to the graph. It reports error enqueue_many  ops has been marked as not fetchable. So I change my code to  return the queuerunner object. It reports errors "Expected binary or unicode string, got <tensorflow.python.training.queue_runner_impl.QueueRunner". How did you solve the problem?

**Comment by vanpersie32** [comment]
@mrry  hi mrry. I have encountered similar problem.  I add queue runner to collection using tf.train.add_queue_runner function  inside the conditional execution context. According to your reply , I should return the operations from true-branch and false-branch function. So I try to return the enqueue operations and the equeue from true-branch and false-branch function. But when I return the queue, tensorflow reports errors: "Expected binary or unicode string, got queue"

**Comment by cuevas1208** [comment]
I ran into the same error "marked as not fetchable" at tf.cond(predicate, lambda: op, lambda: tf.no_op()). The got it around by setting the identity for the opp after  tf.cond()

`tf.identity(self.object_inside_op, name="boxes")`

**Comment by nishantb21** [comment]
Hi, 

I have run into a problem which I think is relevant to this issue. I want to try to reference a pythonic list by evaluating a tensor inside a while loop like so. However, it throws a not fetchable error. I use the following piece of code:
```
import tensorflow as tf

class A():
    def __init__(self):
        self.lst = [1, 2, 3]
        self.sess = tf.Session()
        self.total_length = tf.constant(len(self.lst))

    def loop(self, i):
        pr = tf.print(i)
        current_value = self.lst[i.eval(session=self.sess)]
        with tf.control_dependencies([pr]):
            i = tf.add(i, 1)
        return [i]

    def cond(self, i):
        return tf.less(i, self.total_length) 

    def run(self):
        i = tf.constant(0)
        while_op = tf.while_loop(self.cond, self.loop, [i])
        final_i = self.sess.run(while_op)


if __name__ == "__main__":
    obj = A()
    obj.run()
```

It throws the following error:
```
ValueError: Operation 'while/Identity_1' has been marked as not fetchable.
```
Am I doing something wrong?

**Comment by Jemmagu** [comment]
@mrry excuse me, so if i really want to fetch op created in the tf,while_loop (eg: use tf.eval() to transform a tensor to numpy), what should i do? is that possible? really appreciate your kindness:)

**Comment by mrry** [comment]
@nishantb21 That appears to be a bug. Can you please open a new issue with your example code and someone will take a look?

@Jemmagu No, it is not possible to fetch the result of an op created inside the while loop's body, because the body might execute 0 or more times, based on the loop condition. To get a value out of the loop, you need to return it from the body function (as one of the loop variables), and its final value after all the iterations will be returned from `tf.while_loop()`.

**Comment by erilyth** [comment]
I have a usecase where I haven't been able to work around with tf.no_ops within cond branches.

I run an eval pipeline where I run metric update ops per batch based on a condition on the inputs. Then I run a final metric op to retrieve the metrics.

Pseudocode:
iter = tf.data.Iterator on a dictionary ['input': float32, 'type' string]

def compute_metrics(x):
  _, metric_op = tf.metrics.mean(x, metrics_collections=[tf.GraphKeys.METRIC_VARIABLES], name='my_mean_op')
  return metric_op

update_ops = tf.cond(condition on iter['type'], lambda: compute_metrics(iter['input']), lambda: tf.no_op())
metric_ops = Find the op by name from tf.get_collection(tf.GraphKeys.METRIC_VARIABLES)

I then run sess.run(update_ops) until the iterator is exhausted and then finally do sess.run(metric_ops) to obtain the final metrics.

The issue here is that most tf.metric.xyz methods return both metric and update ops but in my case I want a conditional update but an always present value op which is initialized to 0 in tf.metric.xyz. Any suggestions on how to go about this?
